remill/BC/DeadStoreEliminator.cpp

```diff
@@ -0,0 +1,768 @@
+        auto succ_block_it = successors(block);
+        if (succ_block_it.begin() == succ_block_it.end()) {
+          curr_wl.push_back(block);
+        }
+      }
+    }
+
+// Visit the basic blocks in the worklist and update the block_map.
+void LiveSetBlockVisitor::Visit(void) {
+  // if any visit makes progress, continue the loop
+  while (!curr_wl.empty()) {
+    LOG(INFO) << "LSBV: " << curr_wl.size() << " blocks in worklist";
+    for (auto block : curr_wl) {
+      if (VisitBlock(block)) {
+        // Updates have been made
+        // Add predecessors to next rotation
+        for (llvm::BasicBlock *pred : predecessors(block)) {
+          next_wl.push_back(pred);
+        }
+      }
+    }
+    // update curr_wl to the next set of insts
+    curr_wl.swap(next_wl);
+    // reset upcoming insts
+    next_wl.clear();
+  }
+}
+
+// Return true if the call accesses the state struct.
+// If a single slot is accessed, set the slotn to that slot.
+bool LiveSetBlockVisitor::CallAccessesState(llvm::CallInst *call, uint64_t *slotn) {
+  slotn = nullptr;
+  for (auto &arg_it : call->arg_operands()) {
+    auto arg = arg_it->stripPointerCasts();
+    if (val_to_offset.count(arg)) {
+      auto offset = val_to_offset.at(arg);
+      if (offset != 0) {
+        // if we access a single non-zero offset, mark just that offset
+        *slotn = state_slots[offset].i;
+      }
+      return true;
+    }
+  }
+  return false;
+}
+
+bool LiveSetBlockVisitor::VisitBlock(llvm::BasicBlock *B) {
+  LiveSet live;
+  for (auto inst_it = B->rbegin(); inst_it != B->rend(); ++inst_it) {
+    auto inst = &*inst_it;
+    if (llvm::isa<llvm::ReturnInst>(inst) || llvm::isa<llvm::UnreachableInst>(inst) ||
+        llvm::isa<llvm::IndirectBrInst>(inst) || llvm::isa<llvm::ResumeInst>(inst) ||
+        llvm::isa<llvm::CatchSwitchInst>(inst) || llvm::isa<llvm::CatchReturnInst>(inst) ||
+        llvm::isa<llvm::CleanupReturnInst>(inst)) {
+      // code that we return to or branch to could read out registers
+      // so mark as all live
+      live.set();
+    } else if (llvm::isa<llvm::BranchInst>(inst) || llvm::isa<llvm::SwitchInst>(inst)) {
+      // update live from successors
+      for (llvm::BasicBlock *succ : successors(B)) {
+        // all successors must be default have been visited first
+        // b = U { s | for s in successors(b)}
+        live |= block_map[succ];
+      }
+    } else if (auto call_inst = llvm::dyn_cast<llvm::CallInst>(inst)) {
+      uint64_t slotn = 0;
+      // Alternative 1:
+      // mark as all live if called function has the same type as lifted function
+      // we're not able to see inside other paths, so we can't predict whether or
+      // not the callee might not use any of our slots
+      // Alternative 2:
+      // mark live based on how much of the state struct could be accessed
+      if ((call_inst->getFunctionType() == lifted_func_ty) || 
+          CallAccessesState(call_inst, &slotn)) {
```

Comment:
Lets rename this function to `MarkLiveArgs` or something like that, and now the type of the called function doesn't matter, we *always* do this, and the function becomes sufficient.